High temperature thermistor support

JeanSebastienCarrier-Dyze opened the pull request:
This is a follow up of : #2915 and #2941 (since changes that were made in MarlinDev are brought back in Marlin)

These modifications allow the use of high temperature thermistors in order to handle hot ends temperature up to 500 degrees Celsius.

Because thermistors able to read higher temperature tend to have a hard time reading lower values, two "defines" were added to allow a "preheating time" as well as handle some measure of noise reading (sudden drops for a few frames).

These defines have been added to the default configuration. Using another configuration or commenting these additions produce the same execution than before these changes.

A high temperature compatible thermistor has been added (no 66) to the thermistor table

The tests were taken from MarlinFirmware/Marlin#2066
The hardware was : Board: Arduino MEGA2560 with RAMPS 1.4
_Dyze Design sensor 500°C
HEATER_0_MINTEMP 21
MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED = 5;
MILLISECONDS_PREHEAT_TIME = 20000;_

**TEST 7 - Hotend's thermistor disconnected before startup**
No error is shown. Temperature shown is 20°C, 1°C below the mintemp error.
Since the conditions «is preheating» and «target temperature > 0» is not met, the min_temp error is not shown as it would normally be.

**TEST 9 - Hotend's thermistor disconnected after heating begins AND before (preheating time) seconds elapsed**
Heater remains ON for the duration of "preheating time". Turns OFF once the duration is elapsed.

**TEST 11 - Hotend's thermistor disconnected after heating begins AND after (preheating time) seconds elapsed**
Min_temp error, heater is disabled.

**TEST 13 - Hotend's thermistor disconnected after target temperature reached.**
Min_temp error, heater is disabled.

JeanSebastienCarrier-Dyze committed:
Add high temperature thermistors support

JeanSebastienCarrier-Dyze committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into High-Temperature-Thermistor-Support

JeanSebastienCarrier-Dyze committed:
Comment added variables by default

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -161,6 +161,9 @@ int Temperature::maxttemp_raw[HOTENDS] = ARRAY_BY_HOTENDS(HEATER_0_RAW_HI_TEMP ,
 int Temperature::maxttemp[HOTENDS] = ARRAY_BY_HOTENDS1(16383);
 
+int Temperature::consecutive_low_temperature_error[HOTENDS] = ARRAY_BY_HOTENDS(0, 0, 0, 0);
+unsigned long Temperature::preheatStartTime[HOTENDS] = ARRAY_BY_HOTENDS(0, 0, 0, 0);
```
For arrays initialized to zero, you can just use `{ 0 }` rather than these macros.

thinkyhead commented on Marlin/temperature.h:
```diff
@@ -220,6 +223,18 @@ class Temperature {
+    }
+
+    static void start_preheat_time(int hotend) { preheatStartTime[hotend] = millis(); }
+    static void reset_preheat_time(int hotend) { preheatStartTime[hotend] = 0; }
```
We prefer ending-times, as they require less computation per loop.
And we have some macros for this. So this would be better:

``` cpp
// preheating functions
static bool is_preheating(int hotend) {
  return
    #ifdef MILLISECONDS_PREHEAT_TIME
      preheatEndTime[hotend] && PENDING(millis(), preheatEndTime[hotend])
    #else
      false
    #endif
  ;
}

static void start_preheat_time(uint8_t hotend) { preheatEndTime[hotend] = millis() + MILLISECONDS_PREHEAT_TIME; }
static void reset_preheat_time(uint8_t hotend) { preheatEndTime[hotend] = 0; }
```

thinkyhead commented on Marlin/thermistortables.h:
```diff
@@ -979,6 +979,40 @@ const short temptable_60[][2] PROGMEM = {
 };
 #endif
+
+#if (THERMISTORHEATER_0 == 66) || (THERMISTORHEATER_1 == 66) || (THERMISTORHEATER_2 == 66) || (THERMISTORHEATER_3 == 66) || (THERMISTORBED == 66)
```
Thermistor 66 should be added to the "Thermal Settings" section in `Configuration.h` where it lists all the thermistor numbers.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1756,7 +1779,17 @@ void Temperature::isr() {
+      if (minttemp_raw[2] GE2 current_temperature_raw[2] && !is_preheating(2) && target_temperature[2] > 0.0f) {
+        #ifdef MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED
+          ++(consecutive_low_temperature_error[2]);
+          if (consecutive_low_temperature_error[2] >= MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED)
```
Might as well increment this in the comparison line…
```cpp
if (++consecutive_low_temperature_error[2] >= MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED)
```

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1746,7 +1759,17 @@ void Temperature::isr() {
+      if (minttemp_raw[1] GE1 current_temperature_raw[1] && !is_preheating(1) && target_temperature[1] > 0.0f) {
+        #ifdef MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED
+          ++(consecutive_low_temperature_error[1]);
+          if (consecutive_low_temperature_error[1] >= MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED)
```
Might as well increment these in the comparison lines…
```cpp
if (++consecutive_low_temperature_error[1] >= MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED)
```

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1736,7 +1739,17 @@ void Temperature::isr() {
+      if (minttemp_raw[0] GE0 current_temperature_raw[0] && !is_preheating(0) && target_temperature[0] > 0.0f) {
+        #ifdef MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED
+          ++(consecutive_low_temperature_error[0]);
+          if (consecutive_low_temperature_error[0] >= MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED)
```
Might as well increment these in the comparison lines…

``` cpp
if (++consecutive_low_temperature_error[0] >= MAX_CONSECUTIVE_LOW_TEMPERATURE_ERROR_ALLOWED)
```

Grogyan commented:
I'd prefer seeing people use a PT100 or thermocouple instead of these "high temperature" thermistors

woodencase01 commented:
@Grogyan It is true that the best case would be to use the proper amplification circuit with a platinum or thermocouple sensor, but I prefer a drop-in remplacement with minimum hardware add-on and changes.

All sensors have their advantages and disadvantages. Since it is possible to do a little trick on the firmware to go around the weakness of the high temperature thermistor ( low resolution below 40°C ) , it makes this solution more than interesting;
- Better resolution than an amplified PT100 from 100°C to 400°C (Ultimaker circuit Op-Amp)
- Better resolution than K type thermocouple from 140°C to 325°C (using AD595)
- No add-on circuit required
- Faster response than PT100
- Very stable

I think these sensors have been underrated due to their limitation at low temperature and I think it is very well suited for 3D printer application.

Grogyan commented:
@woodencase01 laziness usually wins out every time.  Why go out and buy a hammer to tap something in place, when a screwdriver works just as well.  
Not really a good analogy. but I reckon you get my drift.
I hope you realize that the PT100  and a AD595 produce a voltage which gets fed into the Arduino ADC which is 10bit (Arduino 2560), normal thermistors also have to utilize the same 10bit ADC channel therefore the resolution is the same.
The MAX6675 uses a on-chip 12bit ADC and the MAX31855 uses a 14bit on-chip ADC therefore higher resolution is available.

I do agree though that Marlin should support a variety of thermistors in the meantime

woodencase01 commented:
@Grogyan Thanks for your reply! I didn't know the MAX6675 and MAX31855 at all, very interesting.

Regarding both the PT100 and AD595, they end up having a different resolution because they use a different portion in the 5V ADC range at the output. The amplified PT100 has 1.1V at 0°C up to 2.65V at 400°C. 
The AD595 output 0V at 0°C up to 3V at 300°C.
The voltage divider with the high temperature thermistor with a 4.7kΩ pullup end up showing .025V et 500°C up to 4.92V at 20°C.

I'm currently writing a blog about temperature sensors for 3D printers, I'll keep you updated :D

JeanSebastienCarrier-Dyze committed:
Rework High-Temp Support to follow Marlin's guideline

- Added thermistor 66 in configuration.h
- Use ending time and macro instead of start Time for preheat function
- Use { 0 } instead of ARRAY_BY_HOTENDS macro  to initialize static
members when applicable
- Wrapped some variables and functions in #ifdef block where relevant

JeanSebastienCarrier-Dyze committed:
Merge conflicts in thermistortables.h

JeanSebastienCarrier-Dyze commented:
Seems like I did a bad merge with thermistortables.h, will fix this asap

JeanSebastienCarrier-Dyze committed:
Revert "Merge conflicts in thermistortables.h"

This reverts commit 988eaf744e069ed53dd6c8f30fde991e8342e9a0, reversing
changes made to 3033d6d8cf5d44556454ad71c3ad071ac0d2635a.

JeanSebastienCarrier-Dyze committed:
Proper merge of thermistortables.h

JeanSebastienCarrier-Dyze committed:
Revert of "Revert "Merge conflicts in thermistortables.h""

Thermistortables should be properly merged and the branch up to date

thinkyhead commented on Marlin/temperature.h:
```diff
@@ -274,6 +296,12 @@ class Temperature {
+      if (celsius == 0.0f)
+        reset_preheat_time(hotend);
+      else if (target_temperature[hotend] == 0.0f)
+        start_preheat_time(hotend);
```
These 4 lines need indentation.

thinkyhead commented on Marlin/temperature.h:
```diff
@@ -220,6 +228,20 @@ class Temperature {
+      #ifdef MILLISECONDS_PREHEAT_TIME
+        return preheat_end_time[hotend] && PENDING(millis(), preheat_end_time[hotend]);
+      #else
+        return false;
```
You should add `UNUSED(hotend);` above this line.

thinkyhead commented on Marlin/temperature.cpp:
```diff
@@ -1736,7 +1744,16 @@ void Temperature::isr() {
       #endif
       if (current_temperature_raw[0] GE0 maxttemp_raw[0]) max_temp_error(0);
-      if (minttemp_raw[0] GE0 current_temperature_raw[0]) min_temp_error(0);
+      if (minttemp_raw[0] GE0 current_temperature_raw[0] && !is_preheating(0) && target_temperature[0] > 0.0f) {
```
Does the addition of the condition `target_temperature[0] > 0.0f` fix a bug? Should we be doing this already?

JeanSebastienCarrier-Dyze replied:
It's a problem/bug that arise with the use of thermistors that don't offer good reading (or any reading at all) at room temperature.

**is_preheating** is like a "grace period" where you're allowed to blindly heat stuff for a while because you know you should soon be able to get decent readings; but if you're still not reading anything after that grace period, then normal security kicks in.

However, when you just turn on your printer at room temperature, you're not trying to heat anything (so `is_preheating` will return false), but `minttemps_raw[#] GE# current_temperature_raw[#]` will return true, because our thermistor can't read anything below 40°C and everything is turned off before the print can even start. The idea behind `target_temperature[0] > 0.0f` is "If I'm not actually trying to heat anything, don't throw a min_temp_error"

thinkyhead commented on Marlin/Configuration.h:
```diff
@@ -353,6 +354,34 @@
+// the better/safer, and shouldn't need to be more than 30 seconds (30000)
+//#define MILLISECONDS_PREHEAT_TIME 0
+
+//===========================================================================
```
Items added to `Configuration.h` will need to be added to all copies of `Configuration.h` in the project. The rest of them live in the `example_configurations` folder. But perhaps these settings are more suited to `Configuration_adv.h`…?

JeanSebastienCarrier-Dyze committed:
Move new defines to configuration_adv

Moved define for high temp thermistors to configuration_adv.h and copied
the segment for every configuration_adv in example_configurations.

Add UNUSED(hotend) in default case of is_preheating to remove warning

JeanSebastienCarrier-Dyze committed:
Merge remote-tracking branch 'refs/remotes/MarlinFirmware/RCBugFix' into High-Temperature-Thermistor-Support

JeanSebastienCarrier-Dyze commented:
Any news on this? Does something else needs improvement?

thinkyhead commented:
Hi @JeanSebastienCarrier-Dyze Everything looks good to me, although I have a couple of cleanups that I would make before merging. If you could squash and rebase these commits it will help a lot. Otherwise I'm stuck making a completely new PR and your moniker won't be on any of the commits. My version of this PR is now posted at #4244.
